src/runner/centralized_kb_hook.cpp

```diff
@@ -0,0 +1,109 @@
+#include "pch.h"
+#include "centralized_kb_hook.h"
+
+namespace CentralizedKeyboardHook
+{
+    struct HotkeyDescriptor
+    {
+        Hotkey hotkey;
+        std::wstring moduleName;
+        std::function<bool()> action;
+
+        bool operator<(const HotkeyDescriptor& other) const 
+        {
+            return hotkey < other.hotkey;
+        };
+
+    };
+
+    std::multiset<HotkeyDescriptor> hotkeyDescriptors;
+    std::mutex mutex;
+    HHOOK hHook{};
+
+    struct DestroyOnExit
+    {
+        ~DestroyOnExit()
+        {
+            Stop();
+        }
+    } destroyOnExitObj;
+
+    LRESULT CALLBACK KeyboardHookProc(_In_ int nCode, _In_ WPARAM wParam, _In_ LPARAM lParam)
+    {
+        if (nCode < 0 || ((wParam != WM_KEYDOWN) && (wParam != WM_SYSKEYDOWN)))
+        {
+            return CallNextHookEx(hHook, nCode, wParam, lParam);
+        }
+
+        const auto& keyPressInfo = *reinterpret_cast<KBDLLHOOKSTRUCT*>(lParam);
+
+        Hotkey hotkey{
+            .win = (GetAsyncKeyState(VK_LWIN) & 0x8000) || (GetAsyncKeyState(VK_RWIN) & 0x8000),
+            .ctrl = static_cast<bool>(GetAsyncKeyState(VK_CONTROL) & 0x8000),
+            .shift = static_cast<bool>(GetAsyncKeyState(VK_SHIFT) & 0x8000),
+            .alt = static_cast<bool>(GetAsyncKeyState(VK_MENU) & 0x8000),
+            .key = static_cast<unsigned char>(keyPressInfo.vkCode)
+        };
+
+        std::function<bool()> action;
+        {
+            // Hold the lock for the shortest possible duration
+            std::unique_lock lock{ mutex };
+            HotkeyDescriptor dummy{ .hotkey = hotkey };
+            auto it = hotkeyDescriptors.find(dummy);
+            if (it != hotkeyDescriptors.end())
+            {
+                action = it->action;
+            }
+        }
+
+        if (action)
+        {
+            if (action())
+            {
+                return 1;
```

Comment:
We will need to add in some code of this form after executing the hotkey action:

        // After invoking the hotkey send a dummy key to prevent Start Menu from activating
        INPUT dummyEvent[1] = {};
        dummyEvent[0].type = INPUT_KEYBOARD;
        dummyEvent[0].ki.wVk = 0xFF;
        dummyEvent[0].ki.dwFlags = KEYEVENTF_KEYUP;
        SendInput(1, dummyEvent, sizeof(INPUT));

This is needed to prevent a regression of https://github.com/microsoft/PowerToys/issues/4578.  https://github.com/microsoft/PowerToys/pull/4874 is the PR where we added this fix to the HotkeyManager interop class. The issue can be reproduced by setting the hotkey for launcher to Win+Space, and pressing Win(down), Space(down), Win(up), Space(up). So we need a dummy key to prevent Start menu from launching. I think it's better for that code to be here rather than on the launcher hotkey side since it would apply to any module using a Win key based hotkey.

Reply:
Done